File: android/app/src/main/java/com/clipsync/android/sync/Bt1SyncTransport.kt
package com.clipsync.android.sync

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.nio.charset.StandardCharsets
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

/**
 * [SyncTransport] over an established bt1 channel, so the unchanged [SyncEngine] session
 * runs on Bluetooth exactly as it does on a WebSocket. Each v1 text message is exactly one
 * encrypted bt1 frame. bt1 has no wire-level close: a graceful close simply closes the
 * stream, and the peer sees [TransportFrame.Closed]. Any frame violation (bad declared
 * length, failed tag) is fatal per docs/protocol-bt1.md section 5 — the link dies and the
 * reconnect loop owns recovery, mirroring the IP path.
 */
class Bt1SyncTransport(
    private val input: InputStream,
    private val output: OutputStream,
    private val channel: Bt1SecureChannel,
    private val closeLink: () -> Unit,
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
) : SyncTransport {
    private val writeMutex = Mutex()

    @Volatile
    private var disposed = false

    override suspend fun receive(): TransportFrame =
        withContext(ioDispatcher) {
            val payload =
                try {
                    Bt1StreamFraming.readEncryptedPayload(input)
                } catch (_: IOException) {
                    null
                } ?: return@withContext closed()
            // A failed tag means loss, reorder, tampering, or a peer bug: the channel state
            // is unrecoverable, so the link dies and the supervisor redials.
            val plaintext = channel.decryptor.tryDecryptPayload(payload) ?: return@withContext closed()
            TransportFrame.Text(String(plaintext, StandardCharsets.UTF_8))
        }

    override suspend fun send(text: String) {
        if (disposed) {
            throw IOException("bt1 link is closed")
        }
        val frame =
            try {
                channel.encryptor.encryptFrame(text.toByteArray(StandardCharsets.UTF_8))
            } catch (failure: IllegalArgumentException) {
                throw IOException("bt1 frame plaintext is outside the accepted window", failure)
            } catch (failure: IllegalStateException) {
                throw IOException("bt1 send counter is exhausted", failure)
            }
        withContext(ioDispatcher) {
            writeMutex.withLock {
                try {
                    Bt1StreamFraming.writeEncryptedFrame(output, frame)
                } catch (failure: IOException) {
                    dispose()
                    throw failure
                }
            }
        }
    }

    override suspend fun close(
        code: Int,
        reason: String,
    ) {
        // bt1 defines no post-handshake close message; closing the stream is the close.
        dispose()
    }

    override fun dispose() {
        disposed = true
        runCatching { closeLink() }
    }

    private fun closed(): TransportFrame {
        dispose()
        return TransportFrame.Closed
    }
}

File: windows/ClipSync.Peer.Bluetooth/Bt1SyncTransport.cs
using System.Net.WebSockets;
using System.Text;
using ClipSync.Core.Security.Bt1;
using ClipSync.Peer.Transport;

namespace ClipSync.Peer.Bluetooth;

/// <summary>
/// <see cref="ISyncTransport"/> over an established bt1 channel, so the unchanged
/// <see cref="ClipSync.Peer.Sessions.SyncSessionEngine"/> runs on Bluetooth exactly as it
/// does on a WebSocket. Each v1 text message is exactly one encrypted bt1 frame. bt1 has no
/// wire-level close: a graceful close simply closes the stream and the peer sees
/// <see cref="TransportFrame.Closed"/>. Any frame violation (bad declared length, failed
/// tag) is fatal per docs/protocol-bt1.md section 5 — the link dies and the accept loop
/// owns recovery, mirroring the IP path.
/// </summary>
public sealed class Bt1SyncTransport : ISyncTransport
{
    private readonly Stream stream;
    private readonly Bt1FrameEncryptor send;
    private readonly Bt1FrameDecryptor receive;
    private readonly SemaphoreSlim sendLock = new(1, 1);
    private volatile bool disposed;

    public Bt1SyncTransport(Stream stream, Bt1FrameEncryptor send, Bt1FrameDecryptor receive)
    {
        this.stream = stream ?? throw new ArgumentNullException(nameof(stream));
        this.send = send ?? throw new ArgumentNullException(nameof(send));
        this.receive = receive ?? throw new ArgumentNullException(nameof(receive));
    }

    public async ValueTask<TransportFrame> ReceiveAsync(CancellationToken cancellationToken)
    {
        byte[]? payload;
        try
        {
            payload = await Bt1StreamFrames.ReadEncryptedPayloadAsync(stream, cancellationToken).ConfigureAwait(false);
        }
        catch (IOException)
        {
            return new TransportFrame.Closed();
        }
        catch (ObjectDisposedException)
        {
            return new TransportFrame.Closed();
        }

        if (payload is null)
        {
            return new TransportFrame.Closed();
        }

        // A failed tag means loss, reorder, tampering, or a peer bug: the channel state is
        // unrecoverable, so the link reads as closed and the reconnect logic owns recovery.
        if (!receive.TryDecryptPayload(payload, out var plaintext))
        {
            CloseStreamQuietly();
            return new TransportFrame.Closed();
        }

        return new TransportFrame.Text(Encoding.UTF8.GetString(plaintext));
    }

    public async ValueTask SendTextAsync(string payload, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(payload);
        if (disposed)
        {
            throw new IOException("The bt1 link is closed.");
        }

        byte[] frame;
        try
        {
            frame = send.EncryptFrame(Encoding.UTF8.GetBytes(payload));
        }
        catch (ArgumentException failure)
        {
            throw new IOException("The bt1 frame plaintext is outside the accepted window.", failure);
        }
        catch (InvalidOperationException failure)
        {
            throw new IOException("The bt1 send counter is exhausted.", failure);
        }

        await sendLock.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            await Bt1StreamFrames.WriteFrameAsync(stream, frame, cancellationToken).ConfigureAwait(false);
        }
        finally
        {
            sendLock.Release();
        }
    }

    public ValueTask CloseAsync(WebSocketCloseStatus status, string description, CancellationToken cancellationToken)
    {
        // bt1 defines no post-handshake close message; closing the stream is the close.
        CloseStreamQuietly();
        return ValueTask.CompletedTask;
    }

    public ValueTask DisposeAsync()
    {
        if (!disposed)
        {
            disposed = true;
            CloseStreamQuietly();
            send.Dispose();
            receive.Dispose();
            sendLock.Dispose();
        }

        return ValueTask.CompletedTask;
    }

    private void CloseStreamQuietly()
    {
        try
        {
            stream.Dispose();
        }
        catch (IOException)
        {
        }
        catch (ObjectDisposedException)
        {
        }
    }
}
